themes/default/content/blog/pulumi-is-imperative-declarative-imperative/index.md

```diff
@@ -0,0 +1,260 @@
+
+On a first run of `pulumi up` of our example program, our engine will first create the bucket, wait for
+the provisioning to be complete, after which the actual bucket id is passed to the creation of
+all the bucket objects. Since none of the bucket objects depend on other resources, these bucket
+objects can all be provisioned concurrently. The engine will
+
+On a second run, assuming no modifications to our example program, the Pulumi engine will compare the to be model
+with the actual state and conclude that nothing needs to be done.
+
+Let's crank up the number of bucket objects to 11.
+
+```python
+from pulumi_aws import s3
+
+bucket = s3.Bucket('bucket')
+
+for i in range(11): # <- nr of objects increased by 1
+    s3.BucketObject(
+        f'object-{i}',
+        s3.BucketObjectArgs(
+            bucket=bucket.id,
+            key=str(i),
+        )
+    )
+```
+
+The third Pulumi run will find the bucket and the first 10 objects in the state. The only action taken
+now is the creation of the 11th bucket object. This brings the actual state back in sync with your
+intended model.
+
+```sh
+$ pulumi up
+
+Updating (<masked>/dev)
+
+View Live: https://app.pulumi.com/<masked>/blog_code/dev/updates/2
+
+     Type                    Name           Status
+     pulumi:pulumi:Stack     blog_code-dev
+ +   └─ aws:s3:BucketObject  object-10      created
+
+Resources:
+    + 1 created
+    12 unchanged
+
+Duration: 5s
+
+```
+
+Although you create the intended model of your infrastructure with an imperative language, our engine definitely
+processes this in a declarative way.
+
+### Providers
+
+The above examples hopefully made clear that the Pulumi engine calculates a set of actions to declaratively bring
+the actual state in sync with your intended model. But our engine is not the component which knows how to
+talk to all the different APIs of cloud and tool vendors. That's the role of the providers.
+
+The engine and the provider processes are connected with a gRPC connection. This is similar to the connection
+between the language host and the engine, with the only difference of which requests are sent. The API
+that providers expose are of a [CRUD](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete) nature:
+
+* Create: create a new resource
+* Check: read information of an existing resource
+* Update: update an existing resource with modified information
+* Delete: delete an existing resource when no longer needed
+
+The providers receive requests from the engine dependent on the set of actions the engine calculated.
+A provider doesn't know anything about the state, the correlation between resource and so forth.
+
+The nature of the provider API is clearly imperative.
+
+### Summary
+
+If the question ever pops up again whether Pulumi is declarative or imperative, the answer is clearly we are both.
+It is only based on which component of our architecture you are talking about:
+
+* Language host: imperative & declarative
```

Comment:
Should we mention program code here as well, since that's what generally leads to the confusion? The code that we write being itself declarative (even though written in an imperative language)?

Reply:
@cnunciato the reason I put both `imperative` and `declarative` is because we support languages in both categories:

- imperative: JS/TS, Go, Python, ...
- declarative: YAML

If that wasn't clear when you read this, do you know how to clarify this?

Reply:
I'd love it if we could just say "declarative", since `new s3.Bucket("foo")` is as declarative a statement as `{ foo: { type: "aws:s3:Bucket" } }` is -- but if that's too much, maybe this?

```suggestion
* Language host (your program code): declarative and imperative
```

Reply:
Fair enough, and I defer to you of course. To me though, this is exactly where things get squishy and lead to confusion: _Look -- it's imperative because it's JavaScript!_.  But`var foo = new Foo()` is a declaration -- it's a declaration in an imperative language, yes, but the expression itself is declarative in the same way Adam's `make poop` reference is. IMO. 😇